File: RateLimiter2/SlidingWindowLog.js
const RateLimiterStrategy = require('./RateLimiterStrategy');

class SlidingWindowLog extends RateLimiterStrategy {
  limit = 0;
  windowMs = 0;
  requests = new Map();
  windowStart = Date.now();

  constructor(limit, windowMs) {
    super();
    this.limit = limit;
    this.windowMs = windowMs;
    this.windowStart = Date.now();
  }

  isAllowed(ip) {
    let currentTime = Date.now();
    if (currentTime - this.windowStart >= this.windowMs) {
      this.windowStart = currentTime;
    }

    if (!this.requests.has(ip)) {
      this.requests.set(ip, [{ time: currentTime }]);
      return true;
    }

    let oldRequests = this.requests.get(ip);
    oldRequests = oldRequests.filter(
      (req) => currentTime - req.time < this.windowMs,
    );

    if (oldRequests.length < this.limit) {
      this.requests.set(ip, [...oldRequests, { time: currentTime }]);
      return true;
    }

    return false;
  }
}

module.exports = SlidingWindowLog;

File: Go/RateLimiter2-go/sliding_window_log.go
package main

import (
	"sync"
	"time"
)

type slidingWindowLogEntry struct {
	Time int64
}

type SlidingWindowLog struct {
	mu       sync.Mutex
	Limit    int
	WindowMs int64
	Requests map[string][]slidingWindowLogEntry
}

func NewSlidingWindowLog(limit int, windowMs int64) *SlidingWindowLog {
	return &SlidingWindowLog{
		Limit:    limit,
		WindowMs: windowMs,
		Requests: make(map[string][]slidingWindowLogEntry),
	}
}

func (s *SlidingWindowLog) IsAllowed(ip string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	currentTime := time.Now().UnixMilli()
	if _, ok := s.Requests[ip]; !ok {
		s.Requests[ip] = []slidingWindowLogEntry{{Time: currentTime}}
		return true
	}
	var valid []slidingWindowLogEntry
	for _, req := range s.Requests[ip] {
		if currentTime-req.Time < s.WindowMs {
			valid = append(valid, req)
		}
	}
	if len(valid) < s.Limit {
		s.Requests[ip] = append(valid, slidingWindowLogEntry{Time: currentTime})
		return true
	}
	return false
}
